### ISSUE TITLE
'object' is inferred for element type of 'Iterable[Union[*Ts]]' (using for-in)

### ISSUE THREAD

**Issue by pinterior** [microsoft/pyright#4196]
**Describe the bug**

iterating over `Iterable[Union[*Ts]]` using `for ... in` syntax, type `object` is inferred for elemnt type insted of  `Union[*Ts]`.

**To Reproduce**

```python3
#!/usr/bin/env python3

from typing import Union, Iterable, List
from typing_extensions import TypeVarTuple, reveal_type

Ts = TypeVarTuple("Ts")

def g(x: "Iterable[Union[*Ts]]"):
    t = x.__iter__()
    while True:
        try:
            yield t.__next__()
        except StopIteration:
            break

def f(x: "Iterable[Union[*Ts]]"):
    for i in x:
        reveal_type(i)
        yield i

reveal_type(f)
reveal_type(g)

w: List[int]  = [i for i in f([1, 2, 3])]
x: List[Union[int, str]] = [i for i in f([1, "foo"])]
y: List[int]  = [i for i in f((1, 2, 3))]
z: List[Union[int, str]] = [i for i in f((1, "foo"))]
```
```
information: Type of "i" is "object"
information: Type of "f" is "(x: Iterable[Union[*Ts@f]]) -> Generator[object, None, None]"
information: Type of "g" is "(x: Iterable[Union[*Ts@g]]) -> Generator[Union[*Ts@g], None, None]"
error: Expression of type "list[object]" cannot be assigned to declared type "List[int]" (reportGeneralTypeIssues)
error: Expression of type "list[object]" cannot be assigned to declared type "List[int | str]" (reportGeneralTypeIssues)
error: Expression of type "list[object]" cannot be assigned to declared type "List[int]" (reportGeneralTypeIssues)
error: Expression of type "list[object]" cannot be assigned to declared type "List[int | str]" (reportGeneralTypeIssues)
```

**Expected behavior**

`f` has the same type as `g` and no error reported for `w`, `x`, `y` and `z`

**VS Code extension or command-line**

command-line tool 1.1.280

**Additonal Context**

1.1.279 inferred a bit different type for `f`

```
information: Type of "f" is "(x: Iterable[*Ts@f]) -> Generator[Ts@f, None, None]"
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.281, which I just published. It will also be included in a future release of pylance.
